### PR TITLE
Feature/uplift tests and cicd

### PR DESCRIPTION
**Pull request by duttonw**
* Majority code coveraged
* pyright included
* tcp + unix clamav test paths (plus some mocking for edge cases)

### CODE REVIEW

File: .github/workflows/pyright.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+permissions:
+  contents: read
+
+jobs:
+  typecheck:
+    strategy:
+      matrix:
+        include: #ckan-image see https://github.com/ckan/ckan-docker-base, ckan-version controls other image tags
+          - ckan-version: "dev-v2.11"
```
Comment (ThrawnCA) [inline .github/workflows/pyright.yml:18]: Shouldn't this use a tag instead of a branch?

File: ckanext/clamav/config.py [hunk 2]
```diff
@@ -1,15 +1,68 @@
...
+from ckan.exceptions import CkanConfigurationException
+
+
+class SocketTypes:
+    # could be Enum but keeping it system for now
+    UNIX = u'unix'
+    TCP = u'tcp'
+
+
+class CalmAvStatus:
```
Comment (ThrawnCA) [inline ckanext/clamav/config.py]: Calm -> Clam

File: ckanext/clamav/tests/test_clamav_local.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    test_file.seek(0)
+    return FlaskFileStorage(test_file, filename)
+
+
+@pytest.mark.usefixtures(u"clean_db", u"clean_index", u'with_plugins')
+@pytest.mark.ckan_config("ckan.plugins", "clamav")
+@pytest.mark.ckan_config("ckanext.clamav.upload_unscanned", "False")
+@pytest.mark.ckan_config("ckanext.clamav.socket_type", "unix")
+@pytest.mark.ckan_config("ckanext.clamav.socket_path", clamav_socket)
+class TestCalmAvLocalDaemon:
```
Comment (ThrawnCA) [inline ckanext/clamav/tests/test_clamav_local.py]: Calm -> Clam

File: ckanext/clamav/tests/test_clamav_local.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                )
+                pytest.fail("Was not tagged as infected:{}".format(res))  # provide nice message if we did not throw exception
+
+            # Verify ClamAV blocked the file
+            assert "{'virus checker': ['malware has been found. filename: eicar.com.txt, signature: win.test.eicar_hdb-1.']}" in str(excinfo.value).lower(), str(excinfo.value).lower()
+
+
+@pytest.mark.usefixtures(u"clean_db", u"clean_index", u'with_plugins')
+@pytest.mark.ckan_config("ckan.plugins", "clamav")
+class TestCalmAvUnitAgentMisconfiguration:
```
Comment (ThrawnCA) [inline ckanext/clamav/tests/test_clamav_local.py]: Calm -> Clam

File: .github/workflows/pytest.yml [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+          image: redis:7
+    env:
+      CKAN_SQLALCHEMY_URL: postgresql://ckan_default:pass@postgres/ckan_test
+      CKAN_DATASTORE_WRITE_URL: postgresql://datastore_write:pass@postgres/datastore_test
+      CKAN_DATASTORE_READ_URL: postgresql://datastore_read:pass@postgres/datastore_test
+      CKAN_SOLR_URL: http://solr:8983/solr/ckan
+      CKAN_REDIS_URL: redis://redis:6379/1
+      CLAMAV_HOST: clamav
+      CLAMAV_PORT: 3310
+      CALMAV_SOCKET: /tmp/clamd.sock
```
Comment (ThrawnCA) [inline .github/workflows/pytest.yml]: CALM -> CLAM

File: ckanext/clamav/tests/test_clamav_local.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+from werkzeug.datastructures import FileStorage as FlaskFileStorage
+
+clean_string = b"safe file content"
+EICAR_URL = "https://secure.eicar.org/eicar.com.txt"
+EICAR_LOCAL_PATH = Path("tests/eicar.com.txt")
+
+# This is to allow localhost testing or container testing by setting CLAMAV_HOST, default localhost
+clamav_host = os.environ.get('CLAMAV_HOST', 'localhost')
+clamav_port = os.environ.get('CLAMAV_PORT', '3310')
+clamav_socket = os.environ.get('CALMAV_SOCKET', '/var/run/clamav/clamd.ctl')
```
Comment (ThrawnCA) [inline ckanext/clamav/tests/test_clamav_local.py]: CALM -> CLAM

File: ckanext/clamav/tests/test_clamav_tcp.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+from werkzeug.datastructures import FileStorage as FlaskFileStorage
+
+clean_string = b"safe file content"
+EICAR_URL = "https://secure.eicar.org/eicar.com.txt"
+EICAR_LOCAL_PATH = Path("tests/eicar.com.txt")
+
+# This is to allow localhost testing or container testing by setting CLAMAV_HOST, default localhost
+clamav_host = os.environ.get('CLAMAV_HOST', 'localhost')
+clamav_port = os.environ.get('CLAMAV_PORT', '3310')
+clamav_socket = os.environ.get('CALMAV_SOCKET', '/var/run/clamav/clamd.ctl')
```
Comment (ThrawnCA) [inline ckanext/clamav/tests/test_clamav_tcp.py]: CALM -> CLAM

File: ckanext/clamav/tests/test_clamav_tcp.py [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    test_file.write(data)
+    test_file.seek(0)
+    return FlaskFileStorage(test_file, filename)
+
+
+@pytest.mark.usefixtures(u"clean_db", u"clean_index", u'with_plugins')
+@pytest.mark.ckan_config("ckan.plugins", "clamav")
+@pytest.mark.ckan_config("ckanext.clamav.upload_unscanned", "False")
+@pytest.mark.ckan_config("ckanext.clamav.socket_type", "tcp")
+class TestCalmAvTcpAgentMisconfiguration:
```
Comment (ThrawnCA) [inline ckanext/clamav/tests/test_clamav_tcp.py]: Calm -> Clam

File: ckanext/clamav/tests/test_clamav_tcp.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+            pytest.fail("Should have thrown configuration exception")
+
+
+@pytest.mark.usefixtures(u"clean_db", u"clean_index", u'with_plugins')
+@pytest.mark.ckan_config("ckan.plugins", "clamav")
+@pytest.mark.ckan_config("ckanext.clamav.upload_unscanned", "False")
+@pytest.mark.ckan_config("ckanext.clamav.socket_type", "tcp")
+@pytest.mark.ckan_config("ckanext.clamav.tcp.host", clamav_host)
+@pytest.mark.ckan_config("ckanext.clamav.tcp.port", clamav_port)
+class TestCalmAvTcpAgent:
```
Comment (ThrawnCA) [inline ckanext/clamav/tests/test_clamav_tcp.py]: Calm -> Clam

File: ckanext/clamav/tests/test_clamav_tcp.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+            package_id=dataset["id"],
+            url="",
+            upload=create_file(clean_string, 'safe.txt'),
+            name="Clean File",
+        )
+
+        # Check if the resource was created
+        assert "id" in res, "Resource was not created successfully:" + res
+
+    def test_clamav_blocks_infected_file_via_tcp_calmd(self, eicar_file_path):
```
Comment (ThrawnCA) [inline ckanext/clamav/tests/test_clamav_tcp.py]: calm -> clam

File: ckanext/clamav/tests/test_config.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
+import pytest
+
+from ckan.exceptions import CkanConfigurationException
+
+import ckanext.clamav.config as c
+
+
+class TestCalmAVConfig:
```
Comment (ThrawnCA) [inline ckanext/clamav/tests/test_config.py]: Calm -> Clam

File: ckanext/clamav/tests/test_plugin.py [hunk 12]
```diff
@@ -1,53 +1,49 @@
...
+from werkzeug.datastructures import FileStorage as FlaskFileStorage
+
+clean_string = b"safe file content"
+EICAR_URL = "https://secure.eicar.org/eicar.com.txt"
+EICAR_LOCAL_PATH = Path("tests/eicar.com.txt")
+
+# This is to allow localhost testing or container testing by setting CLAMAV_HOST, default localhost
+clamav_host = os.environ.get('CLAMAV_HOST', 'localhost')
+clamav_port = os.environ.get('CLAMAV_PORT', '3310')
+clamav_socket = os.environ.get('CALMAV_SOCKET', '/var/run/clamav/clamd.ctl')
```
Comment (ThrawnCA) [inline ckanext/clamav/tests/test_plugin.py]: CALM -> CLAM

File: ckanext/clamav/utils.py [hunk 13]
```diff
@@ -12,17 +12,18 @@
...
 import ckan.model as model
 import ckan.logic as logic
-import ckan.plugins.toolkit as tk
 from ckan.exceptions import CkanConfigurationException
+from ckan.types import ErrorDict
 
-import ckanext.clamav.config as c
-from ckanext.clamav.adapters import CustomClamdNetworkSocket
+from . import config as c
+from .config import CalmAvStatus
```
Comment (ThrawnCA) [inline ckanext/clamav/utils.py]: Calm -> Clam

File: ckanext/clamav/utils.py [hunk 14]
```diff
@@ -33,41 +34,41 @@ def scan_file_for_viruses(data_dict: dict[str, Any]):
...
+    upload_unscanned: bool = c.upload_unscanned()
 
     file: FileStorage = data_dict["upload"]
     status: str
     signature: Optional[str]
     status, signature = _scan_filestream(file)
     package_id = _get_package_id(data_dict)
 
-    if status == c.CLAMAV_STATUS_ERR_DISABLED:
+    if status == CalmAvStatus.ERR_DISABLE:
```
Comment (ThrawnCA) [inline ckanext/clamav/utils.py]: Calm -> Clam

File: ckanext/clamav/utils.py [hunk 15]
```diff
@@ -33,41 +34,41 @@ def scan_file_for_viruses(data_dict: dict[str, Any]):
...
             raise logic.ValidationError(
                 {
                     "Virus checker": [
-                        "The clamav is disabled. Can't uploade the file. Contact administrator"
+                        "The clamav is disabled. Can't upload the file. Contact administrator"
                     ]
                 }
             )
-    elif status in (c.CLAMAV_STATUS_ERR_FILELIMIT,):
+    elif status in (CalmAvStatus.ERR_FILELIMIT,):
```
Comment (ThrawnCA) [inline ckanext/clamav/utils.py]: Calm -> Clam

File: ckanext/clamav/utils.py [hunk 16]
```diff
@@ -33,41 +34,41 @@ def scan_file_for_viruses(data_dict: dict[str, Any]):
...
         log.warning(signature)
         if upload_unscanned:
             log.info(_get_unscanned_file_message(file, package_id))
         else:
-            raise logic.ValidationError({"Virus checker": [signature]})
-    elif status == c.CLAMAV_STATUS_FOUND:
+            err: ErrorDict = {"Virus checker": [f"{signature}"]}
+            raise logic.ValidationError(err)
+    elif status == CalmAvStatus.FOUND:
```
Comment (ThrawnCA) [inline ckanext/clamav/utils.py]: Calm -> Clam

**Review by ThrawnCA — CHANGES_REQUESTED** [review]

File: .github/workflows/pyright.yml (hunk 1 above)
Comment (duttonw) [reply .github/workflows/pyright.yml:18]: I'm using the head of the 2.xx development branch so we don't need to alter the tag all the time. This is how the docker containers work.

File: .github/workflows/pytest.yml (hunk 5 above)
Comment (duttonw) [reply .github/workflows/pytest.yml]: i seem to have missspelled lots :$

File: .github/workflows/pyright.yml (hunk 1 above)
Comment (ThrawnCA) [reply .github/workflows/pyright.yml:18]: This seems prone to builds breaking due to upstream changes rather than anything we did in this repo. I don't like it.

Comment (duttonw) [reply .github/workflows/pyright.yml:18]: I was having difficulties with the docker container ckan installed version being found correctly

File: .github/workflows/flake8.yml [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+  workflow_dispatch:
+
+permissions:
+  contents: read
+
+jobs:
+  lint:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
```
Comment (ThrawnCA) [inline .github/workflows/flake8.yml]: These actions could be v4 and v5.

File: .github/workflows/pytest.yml [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+
+    - name: Install clamav
+      run: |
+        set -eu
+        ## For Debian Container running as root
+        
+        apt-get update
+        apt-get install -y --no-install-recommends clamav clamav-daemon
+        
+        freshclam
```
Comment (ThrawnCA) [inline .github/workflows/pytest.yml:81]: Is `freshclam` available with the packages we have installed? On Amazon Linux we had to separately install `clamav-update`.

File: README.md [hunk 19]
```diff
@@ -4,6 +4,9 @@
 
 This is a basic example that helps to scan uploaded resources for malwares with clamd library.
 
+Note: This interfaces via the iUploader system and needs to be first but does not return an object allowing
```
Comment (ThrawnCA) [inline README.md]: May be worth mentioning that to make it work, this plugin must come earlier in the configured plugin order than any other `IUploader` implementations.

**Review by ThrawnCA — CHANGES_REQUESTED** [review]

File: README.md (hunk 19 above)
Comment (duttonw) [reply README.md]: do you have wording you want for this, i kinda said it but needs to be clearer

File: .github/workflows/pytest.yml (hunk 18 above)
Comment (duttonw) [reply .github/workflows/pytest.yml:81]: it is and it works. if it breaks, we will update it.

File: README.md (hunk 19 above)
Comment (ThrawnCA) [inline README.md]: Perhaps something like:

`Note: This plugin provides a custom IUploader implementation to intercept and scan uploaded files.
However, the implementation does not return an object, allowing the actual IUploader implementation to take effect. If using an actual custom uploader such as ckanext-s3filestore, 'clamav' needs to come earlier in the configured plugin order.`

File: ckanext/clamav/tests/test_mocked_clamav.py [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+@pytest.mark.ckan_config("ckan.plugins", "clamav")
+@pytest.mark.ckan_config("ckanext.clamav.socket_type", "unix")
+@pytest.mark.ckan_config("ckanext.clamav.socket_path", "/this/is/mocked")
+class TestMockClamAVLimitTesting:
+
+    @pytest.mark.ckan_config("ckanext.clamav.upload_unscanned", "False")
+    def test_clamav_hit_file_limit_error_throw_validation_error(self, eicar_file_path):
+        with patch("ckanext.clamav.utils.ClamdUnixSocket") as mock_unix_socket:
+            # Make the instance that gets returned by ClamdUnixSocket()
+            mock_clamd = MagicMock()
```
Comment (ThrawnCA) [inline ckanext/clamav/tests/test_mocked_clamav.py:38]: Several of these setup steps are repeated between tests and could be extracted.

**Review by ThrawnCA — APPROVED** [review]
